File: app/src/main/java/com/robifr/ledger/ui/createqueue/viewmodel/MakeProductOrderViewModel.java
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createqueue.viewmodel;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import com.robifr.ledger.data.model.ProductModel;
import com.robifr.ledger.data.model.ProductOrderModel;
import com.robifr.ledger.util.CurrencyFormat;
import com.robifr.ledger.util.livedata.SafeLiveData;
import com.robifr.ledger.util.livedata.SafeMutableLiveData;
import java.math.BigDecimal;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Objects;
import java.util.Optional;

public class MakeProductOrderViewModel {
  @NonNull private final CreateQueueViewModel _viewModel;

  @NonNull
  private final SafeMutableLiveData<Optional<ProductModel>> _inputtedProduct =
      new SafeMutableLiveData<>(Optional.empty());

  @NonNull
  private final SafeMutableLiveData<String> _inputtedQuantityText = new SafeMutableLiveData<>("");

  @NonNull
  private final SafeMutableLiveData<String> _inputtedDiscountText = new SafeMutableLiveData<>("");

  @NonNull
  private final SafeMutableLiveData<BigDecimal> _inputtedTotalPrice =
      new SafeMutableLiveData<>(BigDecimal.ZERO);

  @Nullable private ProductOrderModel _productOrderToEdit = null;

  public MakeProductOrderViewModel(@NonNull CreateQueueViewModel viewModel) {
    this._viewModel = Objects.requireNonNull(viewModel);
  }

  @NonNull
  public SafeLiveData<Optional<ProductModel>> inputtedProduct() {
    return this._inputtedProduct;
  }

  @NonNull
  public SafeLiveData<String> inputtedQuantityText() {
    return this._inputtedQuantityText;
  }

  @NonNull
  public SafeLiveData<String> inputtedDiscountText() {
    return this._inputtedDiscountText;
  }

  @NonNull
  public SafeLiveData<BigDecimal> inputtedTotalPrice() {
    return this._inputtedTotalPrice;
  }

  @Nullable
  public ProductOrderModel productOrderToEdit() {
    return this._productOrderToEdit;
  }

  /**
   * Get current inputted product order from any corresponding inputted live data. If any live data
   * is set using {@link MutableLiveData#postValue(Object)}, calling this method may not immediately
   * reflect the latest changes. For accurate results in asynchronous operations, consider calling
   * this method inside {@link Observer}.
   */
  @NonNull
  public ProductOrderModel inputtedProductOrder() {
    final Long id =
        this._productOrderToEdit != null && this._productOrderToEdit.id() != null
            ? this._productOrderToEdit.id()
            : null;
    final Long queueId =
        this._productOrderToEdit != null && this._productOrderToEdit.queueId() != null
            ? this._productOrderToEdit.queueId()
            : null;
    double quantity = 0.0;
    long discount = 0L;

    try {
      if (!this._inputtedQuantityText.getValue().isBlank()) {
        quantity =
            CurrencyFormat.parse(
                    this._inputtedQuantityText.getValue(),
                    AppCompatDelegate.getApplicationLocales().toLanguageTags())
                .stripTrailingZeros()
                .doubleValue();
      }

    } catch (ParseException ignore) {
    }

    try {
      if (!this._inputtedDiscountText.getValue().isBlank()) {
        discount =
            CurrencyFormat.parse(
                    this._inputtedDiscountText.getValue(),
                    AppCompatDelegate.getApplicationLocales().toLanguageTags())
                .longValue();
      }

    } catch (ParseException ignore) {
    }

    return ProductOrderModel.toBuilder()
        .withId(id)
        .withQueueId(queueId)
        .withProductId(this._inputtedProduct.getValue().map(ProductModel::id).orElse(null))
        .withProductName(this._inputtedProduct.getValue().map(ProductModel::name).orElse(null))
        .withProductPrice(this._inputtedProduct.getValue().map(ProductModel::price).orElse(null))
        .withTotalPrice(this._inputtedTotalPrice.getValue())
        .withQuantity(quantity)
        .withDiscount(discount);
  }

  public void onProductChanged(@Nullable ProductModel product) {
    this._inputtedProduct.setValue(Optional.ofNullable(product));
    final ProductOrderModel inputtedProductOrder = this.inputtedProductOrder();
    this.onTotalPriceChanged(
        ProductOrderModel.calculateTotalPrice(
            inputtedProductOrder.productPrice(),
            inputtedProductOrder.quantity(),
            inputtedProductOrder.discount()));
  }

  public void onQuantityTextChanged(@NonNull String quantity) {
    Objects.requireNonNull(quantity);

    this._inputtedQuantityText.setValue(quantity);
    final ProductOrderModel inputtedProductOrder = this.inputtedProductOrder();
    this.onTotalPriceChanged(
        ProductOrderModel.calculateTotalPrice(
            inputtedProductOrder.productPrice(),
            inputtedProductOrder.quantity(),
            inputtedProductOrder.discount()));
  }

  public void onDiscountTextChanged(@NonNull String discount) {
    Objects.requireNonNull(discount);

    this._inputtedDiscountText.setValue(discount);
    final ProductOrderModel inputtedProductOrder = this.inputtedProductOrder();
    this.onTotalPriceChanged(
        ProductOrderModel.calculateTotalPrice(
            inputtedProductOrder.productPrice(),
            inputtedProductOrder.quantity(),
            inputtedProductOrder.discount()));
  }

  public void onTotalPriceChanged(@NonNull BigDecimal totalPrice) {
    Objects.requireNonNull(totalPrice);

    this._inputtedTotalPrice.setValue(totalPrice);
  }

  public void onProductOrderToEditChanged(@NonNull ProductOrderModel productOrder) {
    Objects.requireNonNull(productOrder);

    this._productOrderToEdit = productOrder;
    final String quantity =
        CurrencyFormat.format(
            BigDecimal.valueOf(productOrder.quantity()),
            AppCompatDelegate.getApplicationLocales().toLanguageTags(),
            "");
    final String discount =
        CurrencyFormat.format(
            BigDecimal.valueOf(productOrder.discount()),
            AppCompatDelegate.getApplicationLocales().toLanguageTags());

    this.onProductChanged(productOrder.referencedProduct());
    this.onQuantityTextChanged(quantity);
    this.onDiscountTextChanged(discount);
    this.onTotalPriceChanged(productOrder.totalPrice());
  }

  public void onSave() {
    final ArrayList<ProductOrderModel> productOrders =
        new ArrayList<>(this._viewModel.inputtedProductOrders().getValue());
    final int indexToUpdate = productOrders.indexOf(this._productOrderToEdit);

    // Add as new when there's no product order to update,
    // otherwise update them with the one user inputted.
    if (indexToUpdate == -1) productOrders.add(this.inputtedProductOrder());
    else productOrders.set(indexToUpdate, this.inputtedProductOrder());

    this._viewModel.onProductOrdersChanged(productOrders);
  }

  public void onReset() {
    this._productOrderToEdit = null;
    this._inputtedProduct.setValue(Optional.empty());
    this._inputtedQuantityText.setValue("");
    this._inputtedDiscountText.setValue("");
    this._inputtedTotalPrice.setValue(BigDecimal.ZERO);
  }
}

File: app/src/main/java/com/robifr/ledger/ui/createqueue/viewmodel/MakeProductOrderViewModel.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createqueue.viewmodel

import androidx.appcompat.app.AppCompatDelegate
import com.robifr.ledger.data.model.ProductModel
import com.robifr.ledger.data.model.ProductOrderModel
import com.robifr.ledger.ui.SafeLiveData
import com.robifr.ledger.ui.SafeMutableLiveData
import com.robifr.ledger.util.CurrencyFormat
import java.math.BigDecimal
import java.text.ParseException

class MakeProductOrderViewModel(private val _viewModel: CreateQueueViewModel) {
  private val _uiState: SafeMutableLiveData<MakeProductOrderState> =
      SafeMutableLiveData(
          MakeProductOrderState(
              isDialogShown = false,
              product = null,
              formattedQuantity = "",
              formattedDiscount = "",
              totalPrice = 0.toBigDecimal(),
              productOrderToEdit = null))
  val uiState: SafeLiveData<MakeProductOrderState>
    get() = _uiState

  private val _inputtedProductOrder: ProductOrderModel
    get() {
      var quantity: Double = 0.0
      try {
        if (_uiState.safeValue.formattedQuantity.isNotBlank()) {
          quantity =
              CurrencyFormat.parse(
                      _uiState.safeValue.formattedQuantity,
                      AppCompatDelegate.getApplicationLocales().toLanguageTags())
                  .stripTrailingZeros()
                  .toDouble()
        }
      } catch (_: ParseException) {}

      var discount: Long = 0L
      try {
        if (_uiState.safeValue.formattedDiscount.isNotBlank()) {
          discount =
              CurrencyFormat.parse(
                      _uiState.safeValue.formattedDiscount,
                      AppCompatDelegate.getApplicationLocales().toLanguageTags())
                  .toLong()
        }
      } catch (_: ParseException) {}

      return ProductOrderModel(
          id = _uiState.safeValue.productOrderToEdit?.id,
          queueId = _uiState.safeValue.productOrderToEdit?.queueId,
          productId = _uiState.safeValue.product?.id,
          productName = _uiState.safeValue.product?.name,
          productPrice = _uiState.safeValue.product?.price,
          totalPrice = _uiState.safeValue.totalPrice,
          quantity = quantity,
          discount = discount)
    }

  fun onProductChanged(product: ProductModel?) {
    _uiState.setValue(_uiState.safeValue.copy(product = product))
    with(_inputtedProductOrder) {
      _onTotalPriceChanged(ProductOrderModel.calculateTotalPrice(productPrice, quantity, discount))
    }
  }

  fun onQuantityTextChanged(formattedQuantity: String) {
    _uiState.setValue(_uiState.safeValue.copy(formattedQuantity = formattedQuantity))
    with(_inputtedProductOrder) {
      _onTotalPriceChanged(ProductOrderModel.calculateTotalPrice(productPrice, quantity, discount))
    }
  }

  fun onDiscountTextChanged(formattedDiscount: String) {
    _uiState.setValue(_uiState.safeValue.copy(formattedDiscount = formattedDiscount))
    with(_inputtedProductOrder) {
      _onTotalPriceChanged(ProductOrderModel.calculateTotalPrice(productPrice, quantity, discount))
    }
  }

  fun onShowDialog(productOrder: ProductOrderModel? = null) {
    _uiState.setValue(
        MakeProductOrderState(
            isDialogShown = true,
            product = productOrder?.referencedProduct(),
            formattedQuantity =
                productOrder?.let {
                  CurrencyFormat.format(
                      it.quantity.toBigDecimal(),
                      AppCompatDelegate.getApplicationLocales().toLanguageTags(),
                      "")
                } ?: "",
            formattedDiscount =
                productOrder?.let {
                  CurrencyFormat.format(
                      it.discount.toBigDecimal(),
                      AppCompatDelegate.getApplicationLocales().toLanguageTags())
                } ?: "",
            totalPrice = productOrder?.totalPrice ?: 0.toBigDecimal(),
            productOrderToEdit = productOrder))
  }

  fun onCloseDialog() {
    _uiState.setValue(
        MakeProductOrderState(
            isDialogShown = false,
            product = null,
            formattedQuantity = "",
            formattedDiscount = "",
            totalPrice = 0.toBigDecimal(),
            productOrderToEdit = null))
  }

  fun onSave() {
    _viewModel.onProductOrdersChanged(
        _viewModel.uiState.safeValue.productOrders.toMutableList().apply {
          val indexToUpdate: Int = indexOf(_uiState.safeValue.productOrderToEdit)
          // Add as new when there's no product order to update,
          // otherwise update them with the one user inputted.
          if (indexToUpdate == -1) add(_inputtedProductOrder)
          else set(indexToUpdate, _inputtedProductOrder)
        })
  }

  private fun _onTotalPriceChanged(totalPrice: BigDecimal) {
    _uiState.setValue(_uiState.safeValue.copy(totalPrice = totalPrice))
  }
}
